### PR TITLE
Refactor span extractors and unify forward.

### LINKED ISSUE #5145: Refactoring span extractors to fix wrong span width calculation and unify forward arguments
**Is your feature request related to a problem? Please describe.**
1. The calculated span widths (by `span_ends - span_starts`) in `EndpointSpanExtractor.forward` and `BidirectionalEndpointSpanExtractor.forward` are always 1 less than the real width since the `SpanField` use inclusive indices. (e.g., the width of the span `(2, 2)` will be `2 - 2 = 0` by `span_end - span_start`, but it is 1 actually). Is my understanding correct?

2. The argument `sequence_mask` is missing in `SelfAttentiveSpanExtractor.forward` but exists in `EndpointSpanExtractor.forward` and `BidirectionalEndpointSpanExtractor.forward` (although it is use only in `BidirectionalEndpointSpanExtractor`). The removing of `sequence_mask` in `SelfAttentiveSpanExtractor.forward` brings an arguments mismatching problem in the following code.

```
extractors = [
    EndpointSpanExtractor(), SelfAttentiveSpanExtractor(), 
    BidirectionalEndpointSpanExtractor()
]
embeddings = [
    extractor(sequence_tensor, span_indices, sequence_mask, span_indices_mask)
    for extractor in extractors
]
```

3. Just curious, why `SelfAttentiveSpanExtractor` does not have span width embeddings?

**Describe the solution you'd like**
I suggest adding a `__init__` method to `SpanExtractor` for initiating `span_width_embedding` and other attributes, and adding a `_combine_span_width_embeddings` (maybe) method to `SpanExtractor` to calculate the correct span widths, as well as embed and combine them to the span embeddings.

In this way, `SelfAttentiveSpanExtractor`, `EndpointSpanExtractor` and `BidirectionalEndpointSpanExtractor` only need to call the `_combine_span_width_embeddings` method without writing redundant code in their `__init__` and `forward`.

**Describe alternatives you've considered**
I have finished a part of the refactoring of these span extractors in [this repo](https://github.com/izhx/allennlpadd/blob/main/modules/span_extractors/) and used them in my research.
It would be great if these changes could be merged into allennlp.

**Additional context**
Do you have any ideas and suggestions? I am willing to make a PR.

### PR DESCRIPTION
**Pull request by izhx**
Fixes #5145. Thanks! @dirkgr 

Changes proposed in this pull request:

- added a `SpanExtractorWithSpanWidthEmbedding` class with `__init__`, `forward` and `get_input_dim` methods, putting specific span embedding computations into the abstract mehtod `_embed_spans`.
- ~~added the `__init__` method to `SpanExtractor` for initiating `span_width_embedding` and other attributes.~~
- ~~added the `forward` method to `SpanExtractor` to unify `forward` arguments of all extractors.~~
- ~~added comments for `width = span_end - span_start`.~~
- ~~added an `_embed_spans` method, which need to be implemented to compute span embeddings in different ways,  to `SpanExtractor`.~~
- modified `EndpointSpanExtractor`, `SelfAttentiveSpanExtractor` and `BidirectionalEndpointSpanExtractor` accordingly (change `forward` to `_embed_spans`, Inherit from `SpanExtractorWithSpanWidthEmbedding`).

Now, `SelfAttentiveSpanExtractor` can also embed span widths.

### CODE REVIEW

**Review by ArjunSubramonian — COMMENTED** [review]
This personally looks great to me! Thanks for refactoring. I'm adding @dirkgr to look it over as well.

File: allennlp/modules/span_extractors/span_extractor.py [hunk 1]
```diff
@@ -53,6 +100,42 @@ def forward(
         A tensor of shape `(batch_size, num_spans, embedded_span_size)`,
         where `embedded_span_size` depends on the way spans are represented.
         """
+        # shape (batch_size, num_spans, embedding_dim)
+        span_embeddings = self._embed_spans(
+            sequence_tensor, span_indices, sequence_mask, span_indices_mask
+        )
+        if self._span_width_embedding is not None:
+            # width = end_index - start_index + 1 since `SpanField` use inclusive indices.
+            # But here we do not add 1 beacuse we offen initiate the span width
```
Comment (dirkgr) [inline allennlp/modules/span_extractors/span_extractor.py:109]: Typo

**Review by dirkgr — CHANGES_REQUESTED** [review]
I would like to keep `SpanExtractor` as a purely abstract base class, just in case someone wants to build a totally different span extractor. But it makes sense to have a `SpanExtractorWithSpanWidthEmbedding` class that's basically the class you wrote, and have all the concrete implementations derive from that (the way you already have it). Does that make sense?

Thanks for finding and fixing this!

**Comment by izhx** [comment]
> I would like to keep `SpanExtractor` as a purely abstract base class, just in case someone wants to build a totally different span extractor. But it makes sense to have a `SpanExtractorWithSpanWidthEmbedding` class that's basically the class you wrote, and have all the concrete implementations derive from that (the way you already have it). Does that make sense?
> 
> Thanks for finding and fixing this!

Yeah, I totally agree with you! It's going to be flexible for future developments. 
I have made these changes, Thanks!

File: allennlp/modules/span_extractors/span_extractor.py (hunk 1 above)
Comment (izhx) [reply allennlp/modules/span_extractors/span_extractor.py]: Changed `offen` to `often `. 🤣

**Comment by ArjunSubramonian** [comment]
> > I would like to keep `SpanExtractor` as a purely abstract base class, just in case someone wants to build a totally different span extractor. But it makes sense to have a `SpanExtractorWithSpanWidthEmbedding` class that's basically the class you wrote, and have all the concrete implementations derive from that (the way you already have it). Does that make sense?
> > Thanks for finding and fixing this!
> 
> Yeah, I totally agree with you! It's going to be flexible for future developments.
> I have made these changes, Thanks!

@izhx Thanks for your great work! The changes look fine to me. I would edit the description of the PR to reflect what you reverted. Also, as a heads up, @dirkgr is out the rest of this week, so please give us a week to get back to you.

**Comment by izhx** [comment]
> > > I would like to keep `SpanExtractor` as a purely abstract base class, just in case someone wants to build a totally different span extractor. But it makes sense to have a `SpanExtractorWithSpanWidthEmbedding` class that's basically the class you wrote, and have all the concrete implementations derive from that (the way you already have it). Does that make sense?
> > > Thanks for finding and fixing this!
> > 
> > 
> > Yeah, I totally agree with you! It's going to be flexible for future developments.
> > I have made these changes, Thanks!
> 
> @izhx Thanks for your great work! The changes look fine to me. I would edit the description of the PR to reflect what you reverted. Also, as a heads up, @dirkgr is out the rest of this week, so please give us a week to get back to you.

Thank you! 😄

**Review by dirkgr — APPROVED** [review]
Thank you!

**Pull request merged**